MacrobenchmarkSample/macrobenchmark/src/androidTest/java/com/example/macrobenchmark/NonExportedActivityBenchmark.kt

```diff
@@ -0,0 +1,103 @@
+import androidx.benchmark.macro.FrameTimingMetric
+import androidx.benchmark.macro.junit4.MacrobenchmarkRule
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import androidx.test.filters.LargeTest
+import androidx.test.filters.SdkSuppress
+import androidx.test.platform.app.InstrumentationRegistry
+import androidx.test.uiautomator.By
+import androidx.test.uiautomator.Direction
+import androidx.test.uiautomator.UiDevice
+import androidx.test.uiautomator.Until
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+import java.util.concurrent.TimeUnit
+
+@LargeTest
+@SdkSuppress(minSdkVersion = 29)
+@RunWith(AndroidJUnit4::class)
+class NonExportedActivityBenchmark {
+    @get:Rule
+    val benchmarkRule = MacrobenchmarkRule()
+
+    @Test
+    fun start() {
+        val instrumentation = InstrumentationRegistry.getInstrumentation()
+        val device = UiDevice.getInstance(instrumentation)
+        benchmarkRule.measureRepeated(
+            packageName = PACKAGE_NAME,
+            metrics = listOf(FrameTimingMetric()),
+            // Try switching to different compilation modes to see the effect
+            // it has on frame timing metrics.
+            compilationMode = CompilationMode.None,
+            iterations = 3,
+            setupBlock = {
+                // Before starting to measure, navigate to the UI to be measured
+                val intent = Intent()
+                intent.action = ACTION
+                startActivityAndWait(intent)
+                val launchRecyclerActivity = device.findObject(
+                    By.res(PACKAGE_NAME, LAUNCH_RESOURCE_ID)
+                )
+                launchRecyclerActivity.click()
+                device.waitUntilActivity(RECYCLER_ACTIVITY_CLASS)
+            }
+        ) {
+            val recycler = device.findObject(
+                By.res(
+                    PACKAGE_NAME,
+                    RECYCLER_RESOURCE_ID
+                )
+            )
+            // Set gesture margin to avoid triggering gesture navigation
+            // with input events from automation.
+            recycler.setGestureMargin(device.displayWidth / 5)
+            for (i in 1..10) {
+                recycler.scroll(Direction.DOWN, 2f)
+                device.waitForIdle()
+            }
+            // Mark activity as finished
+            device.pressBack()
+        }
+    }
+
+    companion object {
+        private const val PACKAGE_NAME = "com.example.macrobenchmark.target"
+        private const val RECYCLER_ACTIVITY_CLASS = "$PACKAGE_NAME.NonExportedRecyclerActivity"
+        private const val ACTION = "$PACKAGE_NAME.ACTIVITY_LAUNCHER_ACTIVITY"
+        private const val LAUNCH_RESOURCE_ID = "launchRecyclerActivity"
+        private const val RECYCLER_RESOURCE_ID = "recycler"
+
+        /**
+         * Waits until an [android.app.Activity] with the given `className` is visible.
+         */
+        fun UiDevice.waitUntilActivity(className: String) {
```

Comment:
In my testing, `By.clazz()` only looks for view classes, not activity classes. Can you confirm it's not just hitting timeout?

What you can do is just what we do here: https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:benchmark/macro/src/main/java/androidx/benchmark/macro/MacrobenchmarkScope.kt;l=102;drc=3c2bd027309bfdbd148967952355db805c0d7db7

My understanding is simply by using `Until.hasObject(By.pkg(packageName).depth(0))`, it will wait for an activity to be ready of the package name to be ready to display, so will wait for its inflation to complete.

Reply:
I have a 10 second timeout, and it definitely does not for wait for that long. 😕 

`Until.hasObject(By.pkg(packageName).depth(0))` may not be the right thing given we already have a visible activity with the same package name right ?

Reply:
> given we already have a visible activity

That's what my initial thinking, but in my testing hasObject() would always pause until the in-progress launch completed.


Can you confirm that the activity arg you pass is used? That is, you will wait for the 10 second timeout if you pass a different activity string? I'm pretty sure I tested that behavior, and it didn't work with fully qualified activity names. I tried manually passing a string, as well as querying intent.getComponent().name (or whatever it is).

On that note, be sure to throw if the wait(hasObject(...)) returns false, so there's no ambiguity whether it found the class.

Reply:
Yes, there is a 10 second delay, if i refer to a class that does not exist. So it looks like the name is definitely being used. 

> On that note, be sure to throw if the wait(hasObject(...)) returns false, so there's no ambiguity whether it found the class.
That should happen already, because `findObject(...)` returns `null` and will result in an NPE.